feat(sdp): sdp connection to cilium agent

vipul-21 opened the pull request:
- Connection manager to manage the gRPC connection to cilium agent
- Periodic job for handling the cilium agent connection and observer job for stream receiving the DNS policy rules.
- The `ConnectToAgent` is responsible for connection to the agent. It runs periodically (every 10 seconds). If it is successfully, it marks the connection as healthy for the consumer(streams) to use the connection. It needs a write lock for updating the connection.
- `handleConnEvent` listens (observes) for connection event and if the client is connected to agent, it creates a stream to listen for dns policy rules. (though `handlePolicyStream`). `handlePolicyStream` needs a read lock on the connection to recv state from the cilium agent. On connection error, it `removesConnection` based on the revision it was holding. (`standalone-dns-proxy/pkg/client/client.go`)
- `NotifyOnMsg` is called by the DNS proxy to send the mappings to the agent. This also needs a read lock on the connection and has the similar error handling as that of `removesConnection`.  (`standalone-dns-proxy/pkg/client/client.go`)
- All the access to the connection is controlled via `connectionManager`. 
    - Expose a blocking accessor `getFqdnClientWithRev()` that:
        - Waits (via sync.Cond) until a connection is present.
        - Returns an immutable snapshot: (`fqdnClient, revision`).
    - Provide revision-guarded conditional teardown so stale goroutines cannot close a newer connection. 
    - A RW lock protects read vs write access to the active client (fast readers: stream + Notify paths).
    - A separate mutex + cond var gate the blocking wait for initial (or re) connection.
    - Writers (updateConnection, removeConnection, Close) take the write path, bump a monotonically increasing revision, broadcast the cond, and emit an event.
    - Readers take the read lock only after the cond signals a non‑nil connection; they retain the returned revision to validate teardown attempts.

Fixes: https://github.com/cilium/cilium/issues/30984
CFP: https://github.com/cilium/design-cfps/pull/54


```release-note
feat(sdp): Standalone DNS proxy <> cilium agent connection implementation
```

vipul-21 committed:
feat(sdp): sdp connection to cilium agent

- Connection manager to manage the gRPC connection to cilium agent
- Periodic job for handling the cilium agent connection and observer job
 for stream receiving the DNS policy rules.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

squeed commented on standalone-dns-proxy/cmd/standalonednsproxy.go:
```diff
@@ -57,8 +57,11 @@ func NewStandaloneDNSProxy(params standaloneDNSProxyParams) *StandaloneDNSProxy
-	sdp.connHandler.StartConnection()
+	// watch the connection state and start the DNS proxy once connected
+	for !sdp.connHandler.IsConnected() {
+		sdp.logger.Debug("Waiting for connection to Cilium agent...")
```
Will this not cause log spam?

vipul-21 replied:
In the worst case scenario where cilium agent is never up(for any reason), yes we would see this log spamming the logs line for debug mode. but in that case standalone dns proxy is noob daemonset as it won't be running the dns proxy and just waiting for agent connection.
I can remove this as well if it is a concern.

vipul-21 replied:
Removed for now.

squeed commented on standalone-dns-proxy/cmd/standalonednsproxy.go:
```diff
@@ -57,8 +57,10 @@ func NewStandaloneDNSProxy(params standaloneDNSProxyParams) *StandaloneDNSProxy
 
-	// start the connection handler
-	sdp.connHandler.StartConnection()
+	// watch the connection state and start the DNS proxy once connected
```
This function is in the Hive Start hook, which means it is not allowed to take a long time. Connecting may take an unbounded amount of time, as the agent may be down. You'll need to refactor this to be in its own job.

vipul-21 replied:
I updated it to be a job that watches over the connection.

squeed reviewed (changes requested):
So, the gRPC connection state doesn't work like a "standard" one-and-done HTTP connection. Rather, the gRPC client itself maintains a connection pool and has an internal state machine. I have learned this the **very hard** way.

One immediate effect of this is that connections may drop due to idleness. Additionally, it won't actually try and make a connection until the first RPC is issued. The end result is that you can't try and treat gRPC this way, because it will actively fight you the whole time and cause lots of bugs.

Instead, you create a single client and handle RPC failures in the caller itself.

vipul-21 committed:
Remove deterministic checks for gRPC connection

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

squeed commented on standalone-dns-proxy/pkg/client/client.go:
```diff
@@ -132,33 +169,207 @@ type GRPCClient struct {
+	}
+	defer c.connectInProgress.Store(false)
+
+	conn, err := c.dialClient.Dial(
```
grpc Dial is deprecated and with good reason; I don't recommend building tooling on top of this. Just create the client once and let the internal gRPC connection pooling take care of the work.

Rather, build the client once. Set up a separate job that calls `c.handlePolicyStream()`, and make that able to handle errors. Doing so will save you a lot of pain with fighting with the internal gRPC state machine.

vipul-21 replied:
I agree with you and that's what we are doing. To clarify the `Dial` here is a [custom interface `dialClient` ](https://github.com/cilium/cilium/pull/41507/files#diff-327809d94de017325d86f52ad46e202bd3e75977e47b0364c3d32c9075d01de8R131-R147)that calls the `NewClient` as the recommendation. (Used this so that we can add end to end test for the client<>agent connection).  

```
// defaultDialClient implements dialClient by using grpc.NewClient.
type defaultDialClient struct{}

func (d *defaultDialClient) Dial(target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
	return grpc.NewClient(target, opts...)
}
```
Also mentioned here: https://github.com/cilium/cilium/pull/41507#issuecomment-3388044171

>Rather, build the client once. Set up a separate job that calls c.handlePolicyStream(), and make that able to handle errors. Doing so will save you a lot of pain with fighting with the internal gRPC state machine.

The client is build once unless there is a failure reported by the grpc streams. The `func (c *GRPCClient) handlePolicyStream(ctx context.Context)` and `func (c *GRPCClient) NotifyOnMsg(msg *pb.FQDNMapping)` are updating the connection manager based on the error they receive on the stream. This way the rpc methods failures are controlling whether we create a new client or not.

squeed replied:
> The client is build once unless there is a failure reported by the grpc streams.

But the existing gRPC client will continue to reconnect in the background. This code seems susceptible to race conditions because _the client is always retrying_. We used to have similar code to this, and it consistently lost notifications.

The intended lifecycle of gRPC is quite clear: one client for the lifecycle of the process. I don't see a compelling reason to deviate from this.

vipul-21 replied:
Understood, make sense. Removed that custom connection manager.

squeed commented on standalone-dns-proxy/pkg/client/client.go:
```diff
@@ -132,33 +169,207 @@ type GRPCClient struct {
+
+	// connectInProgress indicates if a connection attempt is currently in progress to prevent overlapping dials
+	// This field is used to ensure that only one connection attempt is made at a time.
+	connectInProgress atomic.Bool
```
I think we can get rid of all of these state variables.

vipul-21 replied:
Want to clarify that this state variables are not checking the grpc state but makes sure that there aren't any concurrent jobs that are triggered in case the last one took more than 10sec(interval of the job). 

I removed it for now, as that scenario of overlapping jobs is unlikely.

vipul-21 committed:
start sdp and allow req on conn failure to agent

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

squeed commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -1042,8 +1045,12 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
-		return
+		// If the error is from the standalone dns proxy and is a connection error from cilium agent,
+		// we should be able to forward the request anyway.
+		if !errors.Is(err, ErrDNSProxyConnection) {
```
This is leaky; the DNS proxy shouldn't care about the details of the error response. Rather, NotifyOnDNSMsg shouldn't return an error.

vipul-21 replied:
Updated it.

squeed commented on standalone-dns-proxy/cmd/standalonednsproxy.go:
```diff
@@ -75,6 +77,30 @@ func (sdp *StandaloneDNSProxy) StartStandaloneDNSProxy() error {
+			if sdp.connHandler.IsConnected() {
+				sdp.logger.Info("Connection to Cilium agent established")
+				// Start the DNS proxy once the connection is established
+				if err := sdp.dnsProxier.Listen(sdp.proxyPort); err != nil {
```
Careful, at this point, you don't have rules. You can't open the proxy until you get the first set of rules.

vipul-21 replied:
You're right, I updated the `boolean` to be `true` once we have the rules added in the proxy so the proxy starts after we have the rules. This does mean in case of no rules, standalone dns proxy will just be a dummy process.

vipul-21 committed:
feat(sdp): Remove connection manager for grpc conn

Since gRPC manages the internal connection state,
we've eliminated the need for a separate connection manager.
Instead, we now create the client at the start of the process
and reuse it for different RPC methods. In the event of a
failure, the RPC method can retry, and the connection should
eventually succeed once it's healthy.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>

squeed commented on standalone-dns-proxy/pkg/client/client.go:
```diff
@@ -175,34 +211,163 @@ type GRPCClient struct {
+// if the policy stream is not already established.
+func (c *GRPCClient) createPolicyStream(ctx context.Context) error {
+	if !c.IsConnected() {
+		go c.handlePolicyStream()
```
Hmm, I'm not sure this needs to be an additional goroutine, since every Job is already a separate goroutine.

vipul-21 replied:
Make sense, removed the go routine.

squeed commented on standalone-dns-proxy/cmd/standalonednsproxy.go:
```diff
@@ -75,6 +77,30 @@ func (sdp *StandaloneDNSProxy) StartStandaloneDNSProxy() error {
 }
 
+// watchConnection watches the connection state
+func (sdp *StandaloneDNSProxy) WatchConnection(ctx context.Context, _ cell.Health) error {
```
Do you need this job? What about a separate WithTrigger job that actually starts the proxy? Then, when the first set of rules is received in `handlePolicyStream()`, trigger that? Or, have the proxy be a job and it waits for a channel to be closed.

But these are minor suggestions, non-blocking.

vipul-21 replied:
I don't see a `WithTrigger` option for one shot job. We need to make it a `Timer`  job to have the trigger. I think that implementation is somewhat similar to what we have now. (let me know if this is incorrect)
Waiting on channel makes sense, will keep it out of this PR and update it.

squeed reviewed (approved):
This looks pretty good. Thanks for working through it with me.

I have some minor suggestions about how to best structure the jobs. It seems they are overly complicated. My suggestion would be something along the lines of a job that watches for rules, and job that starts the proxy. 2 jobs, nice and simple. I won't block this PR for that, though.

vipul-21 committed:
Update the policyStream job to remove extra go routine

Policy rules is a part of a Job which is already a go routine, so
removing creation of extra go routine inside it.

Signed-off-by: Vipul Singh <singhvipul@microsoft.com>
